Remove pthreads and use native PHP calls

sandrokeil opened the pull request:
This removes pthreads so we can use different PHP versions / event store real world tests. /cc @codeliner @martin-schilling

What do you think?

codeliner commented:
LGTM, but not sure if we should remove pthreads completely. Maybe just offer this version as an alternative?
You can no longer run the benchmark without docker and the docker-compose run calls are included in the execution time (not sure how much time is spent by docker to spin up the containers and if this time varies a lot)

prolic commented:
@sandrokeil I am against this change because it completely changes it's purpose. You have already read and write tests available. The purpose of this "real-world-test" is getting some test-data about what happens when all those things occur at the same time (A writes, B reads, C projects and so on). Doing them one AFTER the other is not the same thing.

sandrokeil commented:
Yes, there is some overhead on starting the containers via Docker Compose, but all started almost at the same time because of the `&`. But yes, pthreads is a bit more simultanously. Have we any other ideas / options?

Maybe using an entrypoint script which starts all the stuff in one run? Yes a bash script for native tests is missing. But we can also add a bash script which starts PHP native like this here but without Docker. 

@prolic Is it ok to start multiple PHP processes with concatenated `&` so it’s similar to the pthreads stuff?

prolic commented:
Multiple processes should work, but that requires a bigger refactoring.
Nothing against it, but I don't have time to do this myself now. But I can
do code reviews.

On Dec 23, 2017 2:46 AM, "Sandro Keil" <notifications@github.com> wrote:

> Yes, there is some overhead on starting the containers via Docker Compose,
> but all started almost at the same time because of the &. But yes,
> pthreads is a bit more simultanously. Have we any other ideas / options?
>
> Maybe using an entrypoint script which starts all the stuff in one run?
> Yes a bash script for native tests is missing. But we can also add a bash
> script which starts PHP native like this here but without Docker.
>
> @prolic <https://github.com/prolic> Is it ok to start multiple PHP
> processes with concatenated & so it’s similar to the pthreads stuff?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/event-store-benchmarks/pull/11#issuecomment-353654743>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvDQX74GLWdTAak06jSGvFYMgamXfks5tC_jsgaJpZM4RLMi3>
> .
>

sandrokeil commented:
Alright I will do it

codeliner commented:
cool, sounds good. Really looking forward to see ArangoDB with native driver numbers for the real world test

sandrokeil commented:
This is ready for review /cc @prolic @codeliner What I've made:

- Removed `\Thread` inheritance
- Split benchmark in multiple files (prepare, cleanup, benchmark, projector, writer) to allow parallel calls
- Basic tuning of MySql / MariaDB, tests look now much better
- Optimized output and added writing to log files `/logs`
- Added bash scripts for docker / without docker
- Updated readme

The Docker setup works, maybe @prolic can test for native support of some drivers.

sandrokeil commented:
Ping @prolic @codeliner

prolic commented:
Sorry, will try the next 1-2 days

On Jan 2, 2018 5:49 PM, "Sandro Keil" <notifications@github.com> wrote:

> Ping @prolic <https://github.com/prolic> @codeliner
> <https://github.com/codeliner>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/event-store-benchmarks/pull/11#issuecomment-354731312>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvCNe49hrQGXoopzlFEcPcfrgn_wzks5tGfvDgaJpZM4RLMi3>
> .
>

sandrokeil commented:
If this is merged, I will bring a new PR which uses the [ArangoDB PHP ext](https://github.com/sandrokeil/arangodb-php-driver) and [ArangoDB PHP ext polyfill](https://github.com/sandrokeil/arangodb-php-driver-polyfill) as well. I'm curious about the polyfill performance. @martin-schilling is working hard to migrate the ArangoDB PHP ext to native Zend Engine. So stay tuned ;-)

prolic commented:
I checked the changes locally, what I found:

```
./bench.sh --driver arangodb                           :(

Starting benchmark arangodb!
2018-01-07T16:02:17.067956: mysql: set up event store tables on database event_store_tests
2018-01-07T16:02:18.068631: postgres: set up event store tables on database event_store_tests
2018-01-07T16:02:19.071347: arangodb: set up event store tables on database _system
```

We shouldn't set up all databases, when I try to test arangodb.

During execution of bench_all.sh native, I received this:
```
PHP Fatal error:  Uncaught ArangoDBClient\ClientException: Got a timeout while waiting for the server's response
  thrown in /home/sasa/code/event-store-benchmarks/vendor/triagens/arangodb/lib/ArangoDBClient/Connection.php on line 470
```

also:

```
PHP Fatal error:  Uncaught ArangoDBClient\ServerException: collection not found: event_streams in /home/sasa/code/event-store-benchmarks/vendor/triagens/arangodb/lib/ArangoDBClient/Connection.php:522
```

but tests still go further, it won't stop because of this.

I excluded mariadb locally for now, because it just takes time to run and I wasn't that much interested in its results, so here are the other ones:

```
mysql real world test duration 264 seconds
mysql avg writes 47 events/second
mysql avg reads 94 events/second
```

```
postgres real world test duration 259 seconds
postgres avg writes 48 events/second
postgres avg reads 96 events/second
```

```
arangodb real world test duration 270 seconds
arangodb avg writes 46 events/second
arangodb avg reads 92 events/second
```

So the benchmarks are now showing pretty much same results no matter which database is used. There seems to be a problem with the time measurement.

The README.md file needs to be updated, because it still mentions pthreads and PHP 7.2.

Last thing (and I know this was my fault when initially building this test-lib):

In src/functions.php we have this:

```
return [
        'mysql' => ...,
        'mariadb' => ...,
        'postgres' => ...,
        'arangodb' => ...,
    ];
```

in several places, and when I want to comment out mariadb f.e., I need to do this in multiple places, can we please optimize for this?

sandrokeil commented on bench.sh:
```diff
@@ -27,71 +21,69 @@ while [[ ${1} ]]; do
-    DB_CPU=${DB_CPU}$[COUNTER*2+1]
-    let COUNTER=COUNTER+1
-done
+export DRIVER=${DRIVER}
```
@prolic I've added an export so the driver env variable should be now availabe in PHP. Can you test it please? If this not works, can you try to use `php -dDRIVER=${DRIVER} src`?

prolic replied:
ok, but what about the time measurement problem? All drivers have about same performance testing with your benchmark modifications.

sandrokeil commented:
> in several places, and when I want to comment out mariadb f.e., I need to do this in multiple places, can we please optimize for this?

An idea is to define always the drivers you would like to test, seperated by `,` e.g. `--driver postgres,arangodb` and use an `explode/foreach` at the appropriate places.

sandrokeil commented:
@prolic Now ready to test. You have to update the composer dependencies too. I've got different results via Docker, so I guess the calculation is correct. Postgres and ArangoDB are almost identical on my system but MariaDB and MySQL are much slower.
